### PR TITLE
hark: rework caching to allow for better unread counts, read all by index

### PR DESCRIPTION
**Pull request by liam-fitzgerald**
Reworks the caching so that we keep a `(jug index @da)` as a cache of what indices have unread notifications in what timeboxes. Uses this cache to unify the unread handling and add a `%read-index` poke, which allows us to automatically read notifications when we mark a chat as read. Additionally, adds unread counts to group tiles as seen below.

Also includes assorted bugfixes:
- Cleans up UX of accepting a group invite from the invite (#3924 )
- Does not mark notifications as unread when their content is navigated to (#3920 )
- Fixes ordering and loading of notifications (#3929 )
- Fixes an issue where the color of the transparency on publish note notifications would not adjust with theme
- Enables scrollback on archived notifications

Capture d’écran 2020-11-16 à 12 25 51 PM

### CODE REVIEW

**Comment by liam-fitzgerald** [comment]
cc: @urcades lmk if design has any opinions on presenting unread counts in group tiles

**Comment by urcades** [comment]
@liam-fitzgerald Yo!

So, our latest prototype depicts the home/launch screen as so:

Screen Shot 2020-11-15 at 11 43 55 PM

So, simple change: Let's make these blue. 

What needs to be determined now is how exactly this is set/toggled/configured! As I see it there are essentially three options:

1. "Display total update count on group tile" toggle located in the existing notification preferences:
Screen Shot 2020-11-16 at 12 31 37 AM

2. Display the same toggle in one's "Ship Settings":
Screen Shot 2020-11-16 at 12 50 27 AM

3. Push this update **without** the ability to toggle it: This essentially opts people into receiving visual tile notifications as long as they've subscribed to updates — we can think of this approach as analogous to the blue dot, which shows the moment you've subscribed to updates and receive an update. We could consider hiding the tile notifications as an extension to the "do not disturb" option, but this seems somewhat obscure.

I think of these three options, locating the setting in the notifications interface (option 1) is currently the most **obvious** approach given the settings that already exist in the inbox. It should probably be toggled "on" by default.

cc'ing @g-a-v-i-n here because he's led the work on notifications and likely has an opinion here!

**Comment by urcades** [comment]
Something about approach #3 is also appealing to me, only for the reason that it's one less toggle/settings item for someone to worry about. If you've subscribed to notifications, you receive a tile count.

I can see people who are more calm-inclined to bemoan this, but anyone opting into notifications in the first place probably just wants to set and forget, not fiddle with toggles.

**Comment by liam-fitzgerald** [comment]
worth noting that hark only tracks unread counts for graphs that have been subscribed to, so we'd have 3 effectively by default.

**Comment by urcades** [comment]
That's really good to know, 3) seems that much more appealing now — I'd wait to hear gav on on any configuration thoughts he might have, still!

**Comment by galenwp** [comment]
My take:

I'd like to see unread counts on tiles for groups where I'm not subscribed to notifications, I just think they should be grey instead of blue. There are groups I'm intentionally still in but don't follow closely. It could be nice to let these counts accumulate as a way of passively following groups without having them end up in your inbox.

**Comment by vvisigoth** [comment]
FTR I think that liam is referring to _channels_ that have been subscribed
to. Is that correct @liam-fitzgerald?
—
~poldec-tonteg
http://urbit.org

File: pkg/interface/src/logic/reducers/hark-update.ts [hunk 1]
```diff
@@ -6,19 +6,21 @@ import {
...
+  | "notificationsGraphConfig" 
+  | "notifications"
+  | "notificationsCount"
+  | "archivedNotifications"
+  | "unreads">;
 
 export const HarkReducer = (json: any, state: HarkState) => {
   const data = _.get(json, "harkUpdate", false);
   if (data) {
+    console.log(data);
```
Comment (matildepark) [inline pkg/interface/src/logic/reducers/hark-update.ts:23]: ( ͡° ͜ʖ ͡°)

File: pkg/interface/src/logic/reducers/hark-update.ts [hunk 2]
```diff
@@ -6,19 +6,21 @@ import {
 } from "~/types";
 import { makePatDa } from "~/logic/lib/util";
 import _ from "lodash";
+import {StoreState} from "../store/type";
```
Comment (matildepark) [inline pkg/interface/src/logic/reducers/hark-update.ts:9]: linter asks that we put spaces between the brackets here — maybe start running eslint before committing?

**Review by matildepark — COMMENTED** [review]
JS side reads fine

**Comment by vvisigoth** [comment]
Any ETA on getting this out?

**Comment by matildepark** [comment]
I’m asking L for a review today and testing tomorrow for tomorrow night deploy. I would like a few other fixes in as well for Hark and our graph apps.
—
~haddef-sigwen
https://urbit.org

**Comment by vvisigoth** [comment]
""dope""
—
~poldec-tonteg
http://urbit.org

**Comment by urcades** [comment]
Very Dope.

Quick note: 

We had a design review yesterday which touched on the above thoughts I posted for tile notifications.

Some edits to be performed immediately:

Screen Shot 2020-11-17 at 1 22 51 PM

We can opt to revert the color of these tile notifications to gray60/gray, for one.

And we can state plainly that for ease of development, these tile notifications are supplied as described in approach 3) stated above, and repeated + modified for clarity here:

User flow: 

1. I join a group, and its tile appears in my home screen. As long as I am unsubscribed to any channels within the group, I see no dot notification, and I see no tile notifications:

Screen Shot 2020-11-17 at 1 30 30 PM

2. The moment I subscribe to a channel within the group, I am asked if I'd like to subscribe to updates from the channel:

Screen Shot 2020-11-17 at 1 29 24 PM

3. If I keep the box unchecked, my state will be the same as Step 1 above. If I check the box, any new posts to the channel will result in items a) populating my updates inbox, b) triggering the appearance of the blue dot, and c) incrementing the tile notification by the amount of updates received

From here on out, the "Home Screen" can be used to quickly ascertain activity occurring in all groups you have joined, whose channels you have subscribed to.

**Comment by urcades** [comment]
A near-term consideration here links to the statement @galenwp made above, which is to possibly introduce introduce the capability to display the TOTAL NUMBER of ALL channel updates within a group on a tile, while still only subscribing to a subset of this TOTAL NUMBER, this subset of which is sent to the Inbox.

**Comment by matildepark** [comment]
@liam-fitzgerald This PR will also want to include the restoration of "sorting by last updated" behaviour in the sidebars for graph resources, if it doesn't already...

**Review by tacryt-socryp — APPROVED** [review]
LGTM, discussed with Liam

**Comment by matildepark** [comment]
Conflict on this btw.

**Comment by matildepark** [comment]
Good to merge?

**Comment by liam-fitzgerald** [comment]
yep!

**Pull request merged**